freebsd/FreeBSDProcessTable.c

```diff
@@ -248,14 +248,46 @@ void ProcessTable_goThroughEntries(ProcessTable* super) {
 
       proc->priority = kproc->ki_pri.pri_level - PZERO;
 
-      if (String_eq("intr", kproc->ki_comm) && (kproc->ki_flag & P_SYSTEM)) {
-         proc->nice = 0; //@etosan: intr kernel process (not thread) has weird nice value
-      } else if (kproc->ki_pri.pri_class == PRI_TIMESHARE) {
-         proc->nice = kproc->ki_nice - NZERO;
-      } else if (PRI_IS_REALTIME(kproc->ki_pri.pri_class)) {
-         proc->nice = PRIO_MIN - 1 - (PRI_MAX_REALTIME - kproc->ki_pri.pri_level);
-      } else {
-         proc->nice = PRIO_MAX + 1 + kproc->ki_pri.pri_level - PRI_MIN_IDLE;
+      switch (PRI_BASE(kproc->ki_pri.pri_class)) {
+         /* Handling of the below is explained in the FreeBSD base system in:
+          *
+          * /usr/src/usr.bin/top/machine.c (function format_nice) */
+         case PRI_ITHD:
+            fp->sched_class = SCHEDCLASS_INTR_THREAD;
+            proc->nice = 0;
+            break;
+
+         case PRI_REALTIME:
+            fp->sched_class = SCHEDCLASS_REALTIME;
+
+            /* Different for KPROCs and user procs */
+            if (kproc->ki_flag & P_KPROC) {
+               proc->nice = kproc->ki_pri.pri_native - PRI_MIN_REALTIME;
+            } else {
+               proc->nice = kproc->ki_pri.pri_user - PRI_MIN_REALTIME;
+            }
+            break;
+
+         case PRI_IDLE:
+            fp->sched_class = SCHEDCLASS_IDLE;
+
+            /* Different for KPROCs and user procs */
+            if (kproc->ki_flag & P_KPROC) {
+               proc->nice = kproc->ki_pri.pri_native - PRI_MIN_IDLE;
+            } else {
+               proc->nice = kproc->ki_pri.pri_user - PRI_MIN_IDLE;
+            }
+            break;
+
+         case PRI_TIMESHARE:
+            fp->sched_class = SCHEDCLASS_TIMESHARE;
+            proc->nice = kproc->ki_nice - NZERO;
+            break;
+
+         default:
+            fp->sched_class = SCHEDCLASS_UNKNOWN;
+            proc->nice = kproc->ki_nice - NZERO;
```

Comment:
I see implementation differences between yours in this PR and the [FreeBSD top source code](https://github.com/freebsd/freebsd-src/blob/main/usr.bin/top/machine.c#L1182) you cited.

Specifically, when the `sched_class` is unknown, FreeBSD `top` simply display "`?`" for the nice value. It doesn't use the `kproc->ki_nice - NZERO` as if it were a regular ("time sharing") process.

Is this behavior difference intentional?

Reply:
I don't see how to represent an unknown nice value since `proc->nice` is an integer in htop. Setting it to zero would be wrong.

Reply:
If you have a sentinel value would it be sensible to display `N/A` in that case?

Reply:
I guess that makes sense.

Reply:
Anything missing here? Any code changes outstanding?

Reply:
I have added a sentinel value for unknown process niceness.